### PR TITLE
Feature/first-in-voxel filter

### PR DESCRIPTION
**Pull request by pravinshinde825**
- Added `first-in-voxel filter`.
      - This filter will do a pointcloud downsizing based on voxel-grid. 
      - One of the limitation to the `voxelcenternearestneibour` filter is it cannot run in streamable mode. To overcome the limitation, I wrote first-in-voxel filter. This filter will only accept the first point entering in the voxel and mark the voxel as populated, After that any point entering in the same voxel will be ignored/skipped.

### CODE REVIEW

**Comment by nicolas-chaulet** [comment]
> I guess I'm not convinced the center approach couldn't be made streamable (see comments in #2629), but if this works for you then I have no objections. I can take a closer look at the code today.

The voxel centre is streamable and could be an option to this filter I guess. The one that is not streamable is the voxel centroid :)

File: filters/FirstInVoxelFilter.cpp [hunk 1]
```diff
@@ -73,24 +73,53 @@ PointViewSet FirstInVoxelFilter::run(PointViewPtr view)
...
-    size_t const h3(std::hash<int>{}(static_cast<uint32_t>(POSITIVE_VALUE((gz - m_Bounds.minz)))));
-    size_t key = std::hash<size_t>{}(HASH_COMBINE((HASH_COMBINE(h1, h2)), h3));
+	/*
+	* Calculate the local voxel coordinates for incoming point, Using the Pivot voxel.
+	* Generate a combined hash key from local coordinates.
+	*/
+    size_t const hx(std::hash<int>{}(std::floor(gx - m_Bounds.minx)));
+    size_t const hy(std::hash<int>{}(std::floor(gy - m_Bounds.miny)));
+    size_t const hz(std::hash<int>{}(std::floor(gz - m_Bounds.minz)));
+    size_t key = std::hash<size_t>{}(HASH_COMBINE((HASH_COMBINE(hx, hy)), hz));
```
[the remark this answers is not in the document]
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Updated to use `std::tuple`.

File: filters/FirstInVoxelFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "FirstInVoxelFilter.hpp"
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{"filters.first-in-voxel",
```
[the remark this answers is not in the document]
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Done!!

File: filters/FirstInVoxelFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+CREATE_STATIC_STAGE(FirstInVoxelFilter, s_info)
+
+std::string FirstInVoxelFilter::getName() const
+{
+    return s_info.name;
+}
+
+void FirstInVoxelFilter::addArgs(ProgramArgs& args)
+{
+    args.add("cell", "Cell size", m_cell).setPositional();
```
[the remark this answers is not in the document]
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: done!!

File: filters/FirstInVoxelFilter.cpp [hunk 4]
```diff
@@ -73,24 +73,53 @@ PointViewSet FirstInVoxelFilter::run(PointViewPtr view)
...
 
-    m_Bounds.grow(gx, gy, gz);
+	static bool initialized=false;
+    if (!initialized)
+    {
+        /*
+		* Save global coordinates of first incoming point. 
+		* This will act as a Pivot for calculation of local coordinates of the voxels.
+		*/
+		m_Bounds.grow(gx, gy, gz);
```
[the remark this answers is not in the document]
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Changed to array.

File: filters/FirstInVoxelFilter.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+bool FirstInVoxelFilter::voxelize(const PointRef point)
+{
+    double gx = point.getFieldAs<double>(Dimension::Id::X) / m_cell;
+    double gy = point.getFieldAs<double>(Dimension::Id::Y) / m_cell;
+    double gz = point.getFieldAs<double>(Dimension::Id::Z) / m_cell;
+
+    m_Bounds.grow(gx, gy, gz);
+
+    size_t const h1(std::hash<int>{}(static_cast<uint32_t>(POSITIVE_VALUE((gx - m_Bounds.minx)))));
```
[the remark this answers is not in the document]
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Not needed.

**Comment by pravinshinde825** [comment]
@chambbj , Sorry for the delayed response.
I addressed comments and updated code accordingly. Below is a snapshot for output according to latest changes.
Capture

File: filters/FirstInVoxelFilter.hpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    std::string getName() const;
+
+private:
+    double m_cell;
+
+    virtual void addArgs(ProgramArgs& args) override;
+    virtual PointViewSet run(PointViewPtr view) override;
+    bool processOne(PointRef& point) override;
+    bool voxelize(const PointRef point);
+    std::set<std::tuple<int,int,int>> m_populatedVoxels;
```
Comment (nicolas-chaulet) [inline filters/FirstInVoxelFilter.hpp]: @pravinshinde825 I think we would get some nice speed boosts by making that one an `unordered_set`, did you look into that?

Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.hpp]: @nicolas-chaulet , Yes I checked with unordered_set as well. 
- In performance point of view `unordered_set` is a bit faster than `std::set` but not much. 
- In memory point of view I noticed approx 60% more memory usage in `unordered_set` than `set`.
- Also we may need to inject a hash funtion for `std::tuple<int,int,int>`.
So, I think `set` is fine.

File: filters/FirstInVoxelFilter.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "FirstInVoxelFilter.hpp"
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{"filters.firstInVoxel",
+                                     "First Entry Voxel Filter", ""};
```
Comment (chambbj) [inline filters/FirstInVoxelFilter.cpp]: Third (currently empty field) should contain link to the docs, which will be http://pdal.io/stages/filters.first-in-voxel.html as things stand currently, though I think the doc page should be updated to match the name of the filter (having dropped the hyphens now).

File: doc/stages/filters.first-in-voxel.rst [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    :ref:`filters.voxelcenternearestneighbor` offers a similar solution,
+    using
+    the coordinates of the voxel center as the query point in a 3D nearest neighbor search. 
+    The nearest neighbor is then added to the output point cloud, along with any existing dimensions.
+
+Options
+-------------------------------------------------------------------------------
+
+cell
+  Cell size in the ``X``, ``Y``, and ``Z`` dimension.
```
Comment (chambbj) [inline doc/stages/filters.first-in-voxel.rst]: Please indicate the default value for the cell size by adding `[Default: 0.001]` to the end of the description.

File: test/unit/filters/FirstInVoxelFilterTest.cpp [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include <pdal/pdal_test_main.hpp>
+
+#include <pdal/StageFactory.hpp>
+#include<pdal/Reader.hpp>
```
Comment (chambbj) [inline test/unit/filters/FirstInVoxelFilterTest.cpp]: These three lines should have spaces after `#include` for consistency.

**Comment by chambbj** [comment]
@pravinshinde825 Things are looking pretty good I think. I get believable results now. Just a few minor edits requested and then you should be good to go.

**Comment by abellgithub** [comment]
I just wonder what, if any, advantage this has over the existing filter.  Could this logic, along with a check for distance, be added to VoxelCenterNearestNeighbor?  Is there some functional advantage to choosing the first point seen, rather than the one nearest to the center of the voxel?

**Comment by nicolas-chaulet** [comment]
@abellgithub I can't thing of a way to get VoxelCenterNearestNeighbor to work in streaming mode. You need to go through the whole file before you can decide which point to keep. The idea here was to have a voxel downsizing that works in streaming mode. Same filter could be used to pick the centre just like PCL does it (as opposed to the first point)

**Comment by chambbj** [comment]
I think it's as simple as, when you encounter an already initialized voxel, check to see if the distance for the current point is smaller than that of the one that first initialized the voxel, and if it is, update the dimensions/swap the point.

**Comment by nicolas-chaulet** [comment]
If you are streaming, how can you swap? The first point has already been given to the next stage no?

**Comment by chambbj** [comment]
This is true. Sorry.

**Comment by abellgithub** [comment]
Sorry.  For some reason I was thinking of this filter as a writer -- that the voxel set was the output.

File: filters/FirstInVoxelFilter.cpp (hunk 7 above)
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Added link to the doc.

File: doc/stages/filters.first-in-voxel.rst (hunk 8 above)
Comment (pravinshinde825) [reply doc/stages/filters.first-in-voxel.rst]: Added link to the doc.

File: test/unit/filters/FirstInVoxelFilterTest.cpp (hunk 9 above)
Comment (pravinshinde825) [reply test/unit/filters/FirstInVoxelFilterTest.cpp]: Added spaces after `#include`

File: filters/FirstInVoxelFilter.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+{
+    /*
+     * Calculate the voxel coordinates for the incoming point.
+     * gx, gy, gz will be the global coordinates from (0, 0, 0).
+     */
+    int gx = point.getFieldAs<double>(Dimension::Id::X) / m_cell;
+    int gy = point.getFieldAs<double>(Dimension::Id::Y) / m_cell;
+    int gz = point.getFieldAs<double>(Dimension::Id::Z) / m_cell;
+
+    static bool initialized = false;
```
Comment (abellgithub) [inline filters/FirstInVoxelFilter.cpp]: This needs to be moved to be a member variable or multiple instances of this filter will be confused.

File: filters/FirstInVoxelFilter.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        m_pivotVoxel[1] = gy; // Y Coordinate of an Pivot voxel
+        m_pivotVoxel[2] = gz; // Z Coordinate of an Pivot voxel
+        initialized = true;
+    }
+
+    /*
+     * Calculate the local voxel coordinates for incoming point, Using the Pivot
+     * voxel.
+     */
+    auto t=std::make_tuple(gx - m_pivotVoxel[0], gy - m_pivotVoxel[1],
```
Comment (abellgithub) [inline filters/FirstInVoxelFilter.cpp]: Spacing

File: filters/FirstInVoxelFilter.cpp (hunk 10 above)
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Yeah, make sense.
Changed to member variable

File: filters/FirstInVoxelFilter.cpp (hunk 11 above)
Comment (pravinshinde825) [reply filters/FirstInVoxelFilter.cpp]: Done!!

**Comment by pravinshinde825** [comment]
Thanks @nicolas-chaulet for the accurate description.
@chambbj , @abellgithub , I addressed all the comments.

**Pull request merged**